src/common/tusb_fifo.c

```diff
@@ -148,27 +148,34 @@ bool tu_fifo_read(tu_fifo_t* f, void * buffer)
 /******************************************************************************/
 uint16_t tu_fifo_read_n (tu_fifo_t* f, void * buffer, uint16_t count)
 {
-  if( tu_fifo_empty(f) ) return 0;
+  if(tu_fifo_empty(f))
+    return 0;
 
   tu_fifo_lock(f);
 
   /* Limit up to fifo's count */
-  if ( count > f->count ) count = f->count;
-
-  uint8_t* buf8 = (uint8_t*) buffer;
-  uint16_t len = 0;
+  if(count > f->count)
+    count = f->count;
 
-  while (len < count)
+  if(count + f->rd_idx < f->depth)
```

Comment:
Shouldn't this one better be `if(count + f->rd_idx <= f->depth)`?

Then we have to take care that f->rd_idx can get out of valid buffer area. We can increment the `f->rd_idx += count;` here. Same at Line 170 and then write

```
if (f->rd_idx >= f->depth){
  f->rd_idx -= f->depth;
}
```
Just after the if condition to be used with both memcpy cases.

Reply:
Yeap